src/python_testing/TC_EWATERHTRM_1_2.py

```diff
@@ -0,0 +1,158 @@
+
+        endpoint = self.user_params.get("endpoint", 1)
+
+        attributes = Clusters.WaterHeaterMode.Attributes
+
+        self.step(1)
+
+        self.step(2)
+        supported_modes = await self.read_mode_attribute_expect_success(endpoint=endpoint, attribute=attributes.SupportedModes)
+        asserts.assert_greater_equal(len(supported_modes), 2,
+                                     "SupportedModes must have at least 2 entries!")
+        asserts.assert_less_equal(len(supported_modes), 255,
+                                  "SupportedModes must have at most 255 entries!")
+        modes = set([m.mode for m in supported_modes])
+        asserts.assert_equal(len(modes), len(supported_modes),
+                             "SupportedModes must have unique mode values")
+
+        labels = set([m.label for m in supported_modes])
+        asserts.assert_equal(len(labels), len(supported_modes),
+                             "SupportedModes must have unique mode label values")
+
+        # common mode tags
+        commonTags = {0x0: 'Auto',
+                      0x1: 'Quick',
+                      0x2: 'Quiet',
+                      0x3: 'LowNoise',
+                      0x4: 'LowEnergy',
+                      0x5: 'Vacation',
+                      0x6: 'Min',
+                      0x7: 'Max',
+                      0x8: 'Night',
+                      0x9: 'Day'}
+
+        # derived cluster defined tags
+        # kUnknownEnumValue may not be defined
+        try:
+            derivedTags = [tag.value for tag in Clusters.WaterHeaterMode.Enums.ModeTag
+                           if tag is not Clusters.WaterHeaterMode.Enums.ModeTag.kUnknownEnumValue]
+        except AttributeError:
+            derivedTags = Clusters.WaterHeaterMode.Enums.ModeTag
+
+        logging.info("Derived tags: %s" % derivedTags)
+
+        # According to the Mode spec:
+        # At least one entry in the SupportedModes attribute SHALL include the Manual mode tag in the ModeTags field list.
+        # At least one entry in the SupportedModes attribute SHALL include the Off mode tag in the ModeTags field list.
+        # An entry in the SupportedModes attribute that includes one of an Off, Manual, or Timed tag
+        # SHALL NOT also include an additional instance of any one of these tag types.
+        off_present = 0
+        manual_present = 0
+        timed_present = 0
+
+        for m in supported_modes:
+            off_manual_timed_present_in_this_mode = 0
+            for t in m.modeTags:
+                is_mfg = (0x8000 <= t.value and t.value <= 0xBFFF)
+                asserts.assert_true(t.value in commonTags.keys() or t.value in derivedTags or is_mfg,
+                                    "Found a SupportedModes entry with invalid mode tag value!")
+                if t.value == Clusters.WaterHeaterMode.Enums.ModeTag.kOff:
+                    off_present += 1
+                    off_manual_timed_present_in_this_mode += 1
```

Comment:
instead of a separate counter, I wonder if using `sum([off_present, manual_present, timed_present])` could be used in the less_equal assertion to reduce number of variables.

Reply:
`off_manual_timed_present_in_this_mode` is initialised inside the loop after `for m in supported_modes`.
The other variables are initialised outside the loop so I don't think what you are suggesting will work.